rust_src/src/frames.rs

```diff
@@ -691,4 +698,135 @@ pub extern "C" fn get_frame_param(frame: LispFrameRef, prop: LispObject) -> Lisp
     frame.get_param(prop)
 }
 
+/// Perform the switch to frame FRAME.
+///
+/// If FRAME is a switch-frame event `(switch-frame FRAME1)', use
+/// FRAME1 as frame.
+///
+/// If TRACK is non-zero and the frame that currently has the focus
+/// redirects its focus to the selected frame, redirect that focused
+/// frame's focus to FRAME instead.
+///
+/// FOR_DELETION non-zero means that the selected frame is being
+/// deleted, which includes the possibility that the frame's terminal
+/// is dead.
+///
+/// The value of NORECORD is passed as argument to select_window_lisp.
+#[no_mangle]
+pub extern "C" fn do_switch_frame(
+    mut frame: LispObject,
+    _track: bool,
+    for_deletion: bool,
+    norecord: LispObject,
+) -> LispObject {
+    let mut current_frame_ref: LispFrameRef = selected_frame();
+
+    // If `frame` is a switch-frame event, extract the frame we should
+    // switch to.
+    if let Some((event_type, cdr)) = frame.into() {
+        if event_type.eq(Qswitch_frame) {
+            if let Some((event_frame, _)) = cdr.into() {
+                frame = event_frame;
+            }
+        }
+    }
+
+    // This used to check for a live frame, but apparently it's possible for
+    // a switch-frame event to arrive after a frame is no longer live,
+    // especially when deleting the initial frame during startup.
+    if let Some(mut target_frame) = frame.as_live_frame() {
+        if target_frame.eq(&current_frame_ref) {
+            return frame;
+        }
+
+        // If a frame's focus has been redirected toward the currently
+        // selected frame, we should change the redirection to point to the
+        // newly selected frame.  This means that if the focus is redirected
+        // from a minibufferless frame to a surrogate minibuffer frame, we
+        // can use `other-window' to switch between all the frames using
+        // that minibuffer frame, and the focus redirection will follow us
+        // around.
+
+        // Apply it only to the frame we're pointing to.
+        #[cfg(feature = "window-system")]
+        {
+            if _track && target_frame.is_gui_window() {
+                let xfocus = unsafe { x_get_focus_frame(target_frame.as_mut()) };
+
+                if let Some(focus) = xfocus.as_frame().and_then(|f| f.focus_frame.as_frame()) {
+                    if focus.eq(&current_frame_ref)
+                        || target_frame
+                            .minibuffer_window
+                            .eq(current_frame_ref.selected_window)
```

Comment:
Do we need the `NILP (focus)` from the C code ?

Reply:
Yes, we likely do. Which means this code needs to be a little more complex. Which means that `if let Some()` needs to be refactor and broken up.